File: HackerRank/CountingValleys.py
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the countingValleys function below.


def countingValleys(n, s):
    cnt = 0
    height = 0  # 현재 높이
    seaLevel = 0

    for now in s:
        preHeight = height
        if now == 'U':
            height += 1
        elif now == 'D':
            height -= 1

        if height == seaLevel and preHeight < 0:
            cnt += 1

    return cnt


if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(input())

    s = input()

    result = countingValleys(n, s)

    fptr.write(str(result) + '\n')

    fptr.close()

File: HackerRank/CountingValleys.cpp
#include <bits/stdc++.h>

using namespace std;

// Complete the countingValleys function below.
int countingValleys(int n, string s)
{
    int cnt = 0; // 계곡 갯수
    int seaLevel = 0;
    int height = seaLevel;

    for (int i = 0; i < n; i++)
    {
        char now = s[i];
        int preHeight = height; // 이전 높이

        if (now == 'U')
            height++;
        else if (now == 'D')
            height--;

        if (height == seaLevel && preHeight < 0)
            cnt++;
    }

    return cnt;
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    int n;
    cin >> n;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    string s;
    getline(cin, s);

    int result = countingValleys(n, s);

    fout << result << "\n";

    fout.close();

    return 0;
}
